src/v3/src/util/deviceFingerprintingUtils.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { getUserAgent, isWindowsPhone } from './browserUtils';
+
+export const isMessageFromCorrectSource = (iframe: HTMLIFrameElement, event: MessageEvent)
+: boolean => event.source === iframe.contentWindow;
+
+// NOTE: This utility is similar to the DeviceFingerprinting.js file used for V2 authentication flows.
+export const generateDeviceFingerprint = (oktaDomainUrl: string): Promise<string> => {
+  const userAgent = getUserAgent();
+  if (!userAgent) {
+    return Promise.reject(new Error('User agent is not defined'));
+  }
+  if (isWindowsPhone(userAgent)) {
+    return Promise.reject(new Error('Device fingerprint is not supported on Windows phones'));
+  }
+
+  let timeout: NodeJS.Timeout;
+  let iframe: HTMLIFrameElement;
+  let listener: (this: Window, ev: MessageEvent) => void;
+  let msg;
+  const formElement = document.querySelector('form[data-se="o-form"]');
+  const promise = new Promise((resolve, reject) => {
```

Comment:
I'd recommend setting the Promise return type to string. The `return undefined` lines below should then show type errors -- IMO it's true that returning undefined is type-unsafe since the method return type is set to `Promise<string>`.

```suggestion
  const promise = new Promise<string>((resolve, reject) => {
```

The type cast at the bottom `as Promise<string>` basically means to ignore the returned type in the method body.

Reply:
Re-reading the logic here, ignore my previous about `return undefined` and type errors. Still think setting the `Promise<string>` generic is a good idea if it's the expected return type for the method.